### PR TITLE
Multithreading improvements

### PR DESCRIPTION
**Pull request by pmconrad**
Part of https://github.com/bitshares/bitshares-core/issues/1079
~Work in progress, please do not merge yet!~

### CODE REVIEW

**Comment by jmjatlanta** [comment]
The commit comments look awesome. I can't wait to dig in!

**Comment by abitmore** [comment]
Restarted Travis build. It complained that received no output when running `all_tests` in parallel in 10 minutes.

File: src/thread/thread.cpp [hunk 1]
```diff
@@ -62,59 +62,58 @@ namespace fc {
...
-            //assert( !"unhandled exception" );
-            //elog( "Caught unhandled exception %s", boost::current_exception_diagnostic_information().c_str() );
+            if( !p->ready() )
+            {
+               wlog( "unhandled exception" );
+               p->set_exception( std::make_shared<unhandled_exception>( FC_LOG_MESSAGE( warn, "unhandled exception: ${diagnostic}", ("diagnostic",boost::current_exception_diagnostic_information()) ) ) );
+            }
+            else
+            { // possibly shutdown?
+               std::cerr << "unhandled exception in thread '" << name << "'\n";
```
Comment (jmjatlanta) [inline src/thread/thread.cpp:103]: To compile on Ubuntu 18.04, need to include iostream or remove stdout streams

File: tests/thread/parallel_tests.cpp [hunk 2]
```diff
@@ -0,0 +1,340 @@
...
+#include <fc/thread/parallel.hpp>
+#include <fc/time.hpp>
+
+struct thread_config {
+  thread_config() {
+     for( int i = 0; i < boost::unit_test::framework::master_test_suite().argc - 1; ++i )
+        if( !strcmp( boost::unit_test::framework::master_test_suite().argv[i], "--pool-threads" ) )
+        {
+           uint16_t threads = atoi(boost::unit_test::framework::master_test_suite().argv[++i]);
+           std::cout << "Using " << threads << " pool threads\n";
```
Comment (jmjatlanta) [inline tests/thread/parallel_tests.cpp:44]: To compile on Ubuntu 18.04, need to remove stdout streams or include iostream header

**Review by jmjatlanta — CHANGES_REQUESTED** [review]
Looks great. I have not tested on mac or Windows. I see there are some platform-specific areas, so will take a hard look at those within the next few days.

File: src/thread/thread.cpp (hunk 1 above)
Comment (pmconrad) [reply src/thread/thread.cpp:103]: Added. Thanks!

File: tests/thread/parallel_tests.cpp (hunk 2 above)
Comment (pmconrad) [reply tests/thread/parallel_tests.cpp:44]: Added. Thanks!

File: include/fc/thread/parallel.hpp [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+         explicit ticket_guard( boost::atomic<future<void>*>& latch );
+         ~ticket_guard();
+         void wait_for_my_turn();
+      private:
+         promise<void>* my_promise;
+         future<void>*  ticket;
+      };
+
+      friend class ticket_guard;
+      boost::atomic<future<void>*> latch;
```
Comment (jmjatlanta) [inline include/fc/thread/parallel.hpp:63]: need to include boost/atomic/atomic.hpp to compile in Windows.

**Review by jmjatlanta — CHANGES_REQUESTED** [review]
Tests ran successfully in Linux, problems on mac and Windows, but highly likely that the problems are unrelated to the multithreading changes.

File: include/fc/thread/parallel.hpp (hunk 3 above)
Comment (pmconrad) [reply include/fc/thread/parallel.hpp:63]: Added, thanks!

**Review by jmjatlanta — APPROVED** [review]
Tested with Boost 1.67 / Ubuntu 18.04. Reviewed code and all looks good. Thank you!

**Comment by pmconrad** [comment]
Thanks!

**Pull request merged**